prelude/prelude.purs

```diff
@@ -702,6 +706,167 @@ module Prelude
       EQ -> compare xs ys
       other -> other
 
+  -- | The `Bounded` type class represents types that are finite partially
+  -- | ordered sets.
+  -- |
+  -- | Instances should satisfy the following law in addition to the `Ord` laws:
+  -- |
+  -- | - Ordering: `bottom <= a <= top`
+  class (Ord a) <= Bounded a where
+    top :: a
+    bottom :: a
+
+  instance boundedBoolean :: Bounded Boolean where
+    top = true
+    bottom = false
+
+  -- | The `Lattice` type class represents types that are partially ordered
+  -- | sets with a supremum (`sup` or `||`) and infimum (`inf` or `&&`).
+  -- |
+  -- | Instances should satisfy the following laws in addition to the `Ord`
+  -- | laws:
+  -- |
+  -- | - Associativity:
+  -- |   - `a || (b || c) = (a || b) || c`
+  -- |   - `a && (b && c) = (a && b) && c`
+  -- | - Commutativity:
+  -- |   - `a || b = b || a`
+  -- |   - `a && b = b && a`
+  -- | - Absorption:
+  -- |   - `a || (a && b) = a`
+  -- |   - `a && (a || b) = a`
+  -- | - Idempotent:
+  -- |   - `a || a = a`
+  -- |   - `a && a = a`
+  class (Ord a) <= Lattice a where
+    sup :: a -> a -> a
+    inf :: a -> a -> a
+
+  instance latticeBoolean :: Lattice Boolean where
+    sup = boolOr
+    inf = boolAnd
+
+  -- | The `BoundedLattice` type class represents types that are finite
+  -- | lattices.
+  -- |
+  -- | Instances should satisfy the following law in addition to the `Lattice`
+  -- | and `Bounded` laws:
+  -- |
+  -- | - Identity:
+  -- |   - `a || bottom = a`
+  -- |   - `a && top = a`
+  class (Bounded a, Lattice a) <= BoundedLattice a
+
+  instance boundedLatticeBoolean :: BoundedLattice Boolean
+
+  -- | The `ComplementedLattice` type class represents types that are lattices
+  -- | where every member is also complemented.
+  -- |
+  -- | Instances should satisfy the following law in addition to the
+  -- | `BoundedLattice` laws:
+  -- |
+  -- | - Complemented:
+  -- |   - `not a || a == top`
+  -- |   - `not a && a == bottom`
```

Comment:
It has to be complemented _and_ distributive for DeMorgan's doesn't it? So would only apply to `BooleanAlgebra` in this hierarchy, and would be implied by the other laws anyway (I think).